src/SecureStore.HashicorpVault/HashicorpVaultClient.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using UiPath.Orchestrator.Extensibility.SecureStores;
+using VaultSharp;
+using VaultSharp.V1.AuthMethods;
+using VaultSharp.V1.AuthMethods.AppRole;
+using VaultSharp.V1.AuthMethods.Cert;
+using VaultSharp.V1.AuthMethods.LDAP;
+using VaultSharp.V1.AuthMethods.Token;
+using VaultSharp.V1.AuthMethods.UserPass;
+
+namespace UiPath.Orchestrator.Extensions.SecureStores.HashicorpVault
+{
+    public class HashicorpVaultClient : IHashicorpVaultClient
+    {
+        private readonly HashicorpVaultContext _context;
+
+        public HashicorpVaultClient(HashicorpVaultContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<string> GetSecretAsync(string secretName)
+        {
+            var keyVaultClient = GetVaultClient();
+            switch (_context.SecretsEngine)
+            {
+                case SecretsEngine.KeyValueV1:
+                    var kvv1Secret = await keyVaultClient.V1.Secrets.KeyValue.V1.ReadSecretAsync<string>(secretName);
+                    return kvv1Secret.Data;
+                case SecretsEngine.KeyValueV2:
+                    var kvv2Secret = await keyVaultClient.V1.Secrets.KeyValue.V2.ReadSecretAsync<string>(secretName);
+                    return kvv2Secret.Data.Data;
+                case SecretsEngine.Cubbyhole:
+                    var cubbySecret = await keyVaultClient.V1.Secrets.Cubbyhole.ReadSecretAsync(secretName);
+                    return cubbySecret.Data[secretName].ToString();
+                default:
+                    return null;
+            }
+        }
+
+        public async Task<string> SetSecretAsync(string secretName, string secretValue)
+        {
+            var keyVaultClient = GetVaultClient();
+            switch (_context.SecretsEngine)
+            {
+                case SecretsEngine.KeyValueV1:
+                    var kvv1Secret = await keyVaultClient.V1.Secrets.KeyValue.V1.WriteSecretAsync(secretName, secretValue);
+                    return kvv1Secret?.WrapInfo.CreationPath;
+                case SecretsEngine.KeyValueV2:
+                    var kvv2Secret = await keyVaultClient.V1.Secrets.KeyValue.V2.WriteSecretAsync(secretName, secretValue);
+                    return kvv2Secret?.WrapInfo.CreationPath;
+                case SecretsEngine.ActiveDirectory:
+                    throw new NotSupportedException("Writing secrets is not supported for Engine ActiveDirectory");
```

Comment:
The credential store declares itself as read/write, but AD is readonly. How would this work. Is AD even useful?